### PR TITLE
Api 40300 526 bd upload refactor

### PR DESCRIPTION
**Pull request by mchristiansonVA**
## Summary

- *This work is behind a feature toggle (flipper): YES*
- Adds new class DisabilityDocumentService to move document-specific logic for 526 uploads out of the BD class, to reduce the concerns handled by the BD class to simply handling the BD upload given a request body.
- Adds new class DocumentServiceBase to move shared document upload logic for all doc types out of the BD class, such that it can be reused for further refactoring of other doc type uploads (e.g. EWS, POA).
- Adds tests to exercise logic in new classes.

## Related issue(s)

[API-40300](https://jira.devops.va.gov/browse/API-40300)

## Testing done

- [x] *New code is covered by unit tests*
- Prior 526 upload logic used ClaimsApi::BD service methods to handle logic specific to the doc type and generate the appropriate upload body for submission to BD /upload.
- New code uses DisabilityDocumentService and DocumentServiceBase classes to handle logic specific to L123 & L023 doc types and generate the appropriate upload body for submission to BD /upload, using new ClaimsApi::BD service method to simply submit to BD /upload.
- New logic requires the following flippers to be enabled:
Flipper.enable :claims_claim_uploader_use_bd
Flipper.enable ::claims_api_bd_refactor
- Tested 526 v1 and v2 submissions, as well as attachment submissions, and confirmed BD /upload success.

## What areas of the site does it impact?
526 v2 synchronous submissions
526 v2 submissions
526 v2 attachment submissions
526 v1 submissions
526 v1 attachment submissions

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        ClaimsApi::BD.new.upload_document(claim_id: claim.id, doc_type:, body:)
+      end
+
+      private
+
+      ##
+      # Generate form body to upload a document
+      #
+      # @return {parameters, file}
+      def generate_upload_body(claim:, doc_type:, pdf_path:, action:, original_filename: nil)
```
Comment (rockwellwindsor-va) [inline modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb:25]: I'm concerned this method will basically need to be duplicated in the other service files (if I am seeing the direction correctly.

Wondering if there is a way to move this to the service_base perhaps and make it universal enough that we can invoke this one method from the different service files? 

I just feel like this method should still be shared, acknowledging of course that the direction may become clearer with another service in the mix to see.

Comment (FonzMP) [reply modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb:25]: good callout here. just had a sync with Matt and I and we were in agreeance 100%. I like the idea of an intermediary service, somewhat generic, and specific to the document functionality. The original idea was to move to servicebase but with that service, it handles most the root level and generic functionalities. The thought now is to have a DocumentService < ServiceBase that would handle some of these shared, generic document specific methods, and then we'll just have a Service for each form type, hopefully with just a singular "extraction" method to get the values needed to call the rest of the services within the DocumentService, which will be the builders. 

Hopefully that makes sense.

File: modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                             participant_id: nil, tracked_item_ids:)
+      end
+
+      def generate_file_name(veteran_name:, claim_id:, form_name:, original_filename:)
+        if form_name == '526EZ'
+          "#{[veteran_name, claim_id, form_name].compact_blank.join('_')}.pdf"
+        else
+          filename = get_original_supporting_doc_file_name(original_filename)
+          "#{[veteran_name, claim_id, filename].compact_blank.join('_')}.pdf"
+        end
```
Comment (FonzMP) [inline modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb:47]: can this be moved to a more generic class? specifically the `"#{[veteran_name, claim_id, filename].compact_blank.join('_')}.pdf"` being built in to it's own method and we move the logic in this method to choose which of the two `filenames` to use in this method, or does the structure `name, claim, file` change from document type to document type and warrants this uniqueness?

File: modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      # Generate form body to upload a document
+      #
+      # @return {parameters, file}
+      def generate_body(claim:, doc_type:, pdf_path:, original_filename: nil)
+        auth_headers = claim.auth_headers
+        veteran_name = compact_veteran_name(auth_headers['va_eauth_firstName'],
+                                            auth_headers['va_eauth_lastName'])
+        birls_file_number = auth_headers['va_eauth_birlsfilenumber']
+        claim_id = claim.evss_id
+        form_name = '526EZ' if doc_type == 'L122'
```
Comment (FonzMP) [inline modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb:32]: L023's are supported in Claim Establisher - are we prepared for that?

File: modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        base_filename = File.basename(original_filename, file_extension)
+        base_filename[0...-12]
+      end
+
+      def doc_type_to_plain_language(doc_type)
+        case doc_type
+        when 'L122'
+          'claim'
+        else
+          'supporting'
```
Comment (FonzMP) [inline modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb:65]: again, L023. think we need to adjust the :32 - something like the below (not sure of specific word choices here, so don't quote me).
```
form_name = doc_type == 'L122' ? '526EZ' : 'supporting'
```

File: modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb (hunk 2 above)
Comment (mchristiansonVA) [reply modules/claims_api/app/services/claims_api/disability_compensation/disability_document_service.rb]: Yes, good idea. Pushed up changes along those lines, so let me know what you think.

**Review by FonzMP — APPROVED** [review]
:shipit:

**Pull request merged**